### PR TITLE
Replaces product code for driver type

### PR DESCRIPTION
**Pull request by d-loret**
Summary:
Replaced product code configuration parameter with driver type to streamline use of a driver on compatible devices with different product codes. The checks against product codes are pushed down to the implementation of the driver itself. This facilitates the use of a driver among devices that have different product codes but share the same functionality (e.g. revisions of a Beckhoff device that do not change functionality). Additionally, this also allows to have standalone drivers for modalities of a device that are mostly mutually exclusive (e.g. Elmo Platinum drive in nominal mode and Elmo Platinum drive in SIL mode).

Test Plan:
Run tests on:
- [x] Elmo Gold drive
- [x] Elmo Platinum drive
- [x] Network with Elmo motor controller and Beckhoff devices

Reviewers: preston-rogers, alex-brinkman

### CODE REVIEW

File: src/jsd_types.h [hunk 1]
```diff
@@ -29,9 +29,29 @@ extern "C" {
...
+  JSD_DRIVER_TYPE_ATI_FTS,
+  JSD_DRIVER_TYPE_EL3104,
+  JSD_DRIVER_TYPE_EL3202,
+  JSD_DRIVER_TYPE_EL3318,
+  JSD_DRIVER_TYPE_EL3162,
+  JSD_DRIVER_TYPE_EL4102,
+  JSD_DRIVER_TYPE_ILD1900,
+  JSD_DRIVER_TYPE_EPD,
+} jsd_driver_type_t;
+
```
Comment (alex-brinkman) [inline src/jsd_types.h:51]: Nit: Can you order these alpha-numerically?

**Review by alex-brinkman — APPROVED** [review]
Despite my comments, I approve these code changes. Make sure to bump major version in the release since the addition of the driver type in the config struct now breaks backwards compability. 

Great work!

File: src/jsd_types.h (hunk 1 above)
Comment (d-loret) [reply src/jsd_types.h]: Done

File: src/jsd.c [hunk 2]
```diff
@@ -382,10 +382,13 @@ bool jsd_init_all_devices(jsd_t* self) {
...
-                                 slave->eep_id)) {
-      ERROR("User product code (%u) does not match device product code (%u)",
-            self->slave_configs[slave_idx].product_code, slave->eep_id);
+    jsd_driver_type_t driver_type = self->slave_configs[slave_idx].driver_type;
+    if (!jsd_driver_is_compatible_with_product_code(driver_type,
+                                                    slave->eep_id)) {
+      ERROR(
+          "User-specified driver type (%u) is incompatible with the device's "
+          "product code (%u).",
+          driver_type, slave->eep_id);
```
[the remark this answers is not in the document]
Comment (d-loret) [reply src/jsd.c]: Done

**Pull request merged**